# Work Record
#### 2023/12/13(수)
 - 1번 평가 항목인 음정 정확도 평가 로직 작성
#### 2023/12/14(목)
 - 1번 평가 항목인 음정 정확도 평가 로직 수정 및 4번 평가 항목인 빠르기 유사도 평가 로직 작성
 - 12번 평가 항목인 음의 길이(노트의 길이) 평가 로직 일부 작성(개선 필요)
#### 2023/12/20(수)
 - 11번 평가 항목인 페달링 평가 로직 작성
 - 8번 평가 항목인 옥타브 평가 로직 작성
 - 빠르기의 변화 값 추출(5번) 코드 일부 작성
 - 1차 통합 테스트 실시

# 1. 음정 정확도

In [248]:
import mido

# Step 1. 음정 정확도 구하기

def calculate_pitch_accuracy(input_midi_path, target_midi_path):
    """
    인풋 MIDI와 정답 MIDI의 pitch 정확도 계산 - Yeong-Min Ko
    
    Args:
        input_midi: 인풋 MIDI 파일
        target_midi: 타겟 MIDI 파일
        
    Return:
        pitch_accuracy 피치 정확도
    """

    # MIDI 파일 읽기
    input_midi = mido.MidiFile(input_midi_path)
    target_midi = mido.MidiFile(target_midi_path)

    input_notes = set(msg.note for msg in input_midi if msg.type == 'note_on')
    target_notes = set(msg.note for msg in target_midi if msg.type == 'note_on')

    common_notes = input_notes.intersection(target_notes)
    pitch_accuracy = (len(common_notes) / len(input_notes)) * 100

    return pitch_accuracy

def extract_note_info(midi_file_path):
    """
    MIDI 파일에서 노트 정보를 추출
    """
    mid = mido.MidiFile(midi_file_path)

    note_info = []  # 각 노트의 정보를 저장할 리스트

    current_time = 0  # 현재 시간 초기화

    for i, track in enumerate(mid.tracks):
        for msg in track:
            current_time += msg.time  # 현재 시간을 누적

            if msg.type == 'note_on':
                note_start = current_time  # 노트 시작 시간
                note_info.append({'start': note_start, 'end': None, 'note': msg.note})
            elif msg.type == 'note_off':
                note_end = current_time  # 노트 종료 시간
                # 현재 노트 정보에 종료 시간 업데이트
                note_info[-1]['end'] = note_end

    return note_info

def display_note_info(note_info, title):
    """
    노트 정보를 출력
    """
    print(f'\nNote information for {title}:')
    print('-----------------------------------')
    for idx, note in enumerate(note_info):
        print(f'Note {idx + 1}:')
        print(f'Start Time: {note["start"]} ms')
        print(f'End Time: {note["end"]} ms')
        print(f'Note Pitch: {note["note"]}')
        print('-----------------------------------')

def calculate_accuracy_level(accuracy):
    """
    피치 정확도에 따라 Level을 계산하는 함수

    Args:
        accuracy: 피치 정확도

    Return:
        계산된 Level
    """
    level = int(min(10, round(accuracy / 10)))  # 10% 단위로 Level 계산, 최대 Level은 10, 반올림 적용
    return level

# 메인 시작 부분
if __name__ == "__main__":
    # Step 2_1. Mido 라이브러리로 정답 데이터와 같은 인풋 데이터 읽기
    input_midi_path1 = 'Prelude1.mid'
    target_midi_path1 = 'Prelude1.mid'
    accuracy1 = calculate_pitch_accuracy(input_midi_path1, target_midi_path1)

    # Step 2_2. Mido 라이브러리로 정답 데이터와 다른 인풋 데이터 읽기
    input_midi_path2 = 'Prelude1.mid'
    target_midi_path2 = 'Fugue1.mid'
    accuracy2 = calculate_pitch_accuracy(input_midi_path2, target_midi_path2)

    # Step 2_3. Mido 라이브러리로 정답 데이터와 다른 인풋 데이터 읽기
    input_midi_path3 = 'Prelude1.mid'
    target_midi_path3 = 'Fugue1.mid'
    accuracy3 = calculate_pitch_accuracy(input_midi_path3, target_midi_path3)

    # Step 2_4. Mido 라이브러리로 정답 데이터와 다른 인풋 데이터 읽기
    input_midi_path4 = 'Fugue3.mid'
    target_midi_path4 = 'Fugue1.mid'
    accuracy4 = calculate_pitch_accuracy(input_midi_path4, target_midi_path1)

    # Step 3. 각각의 정확도 및 Level 출력
    level1 = calculate_accuracy_level(accuracy1)
    level2 = calculate_accuracy_level(accuracy2)
    level3 = calculate_accuracy_level(accuracy3)
    level4 = calculate_accuracy_level(accuracy4)
    
    print(f'Pitch Accuracy testCase 1: {accuracy1:.2f}% - Level {level1}')
    print(f'Pitch Accuracy testCase 2: {accuracy2:.2f}% - Level {level2}')
    print(f'Pitch Accuracy testCase 3: {accuracy3:.2f}% - Level {level3}')
    print(f'Pitch Accuracy testCase 4: {accuracy4:.2f}% - Level {level4}')
    
    # 각 MIDI 파일의 노트 정보 추출
    #note_info_1 = extract_note_info(midi_file_path_1)
    #note_info_2 = extract_note_info(midi_file_path_2)
    #note_info_4 = extract_note_info(midi_file_path_4)

    # 노트 정보 출력
    # display_note_info(note_info_1, 'Prelude1')
    # display_note_info(note_info_2, 'Fugue1')
    # display_note_info(note_info_4, 'Fugue3')

Pitch Accuracy testCase 1: 100.00% - Level 10
Pitch Accuracy testCase 2: 84.38% - Level 8
Pitch Accuracy testCase 3: 84.38% - Level 8
Pitch Accuracy testCase 4: 68.18% - Level 7


# 2. 셈여림

# 3. 셈여림의 변화

# 4. 곡의 빠르기 유사도 점수 구하기

In [246]:
import mido
from mido import MidiFile

def extract_tempo(midi_file):
    """
    MIDI 파일에서 템포 정보를 추출
    """
    mid = MidiFile(midi_file)

    tempo_list = []  # 각 트랙의 템포를 저장할 리스트

    for i, track in enumerate(mid.tracks):
        for msg in track:
            if msg.type == 'set_tempo':
                # 템포 메시지에서 마이크로초당 박자 속도를 추출
                tempo = mido.tempo2bpm(msg.tempo)
                tempo_list.append(tempo)

    # 템포 정보가 없는 경우 기본값으로 None을 반환
    if not tempo_list:
        return None

    # 전체 트랙의 평균 템포를 계산
    average_tempo = sum(tempo_list) / len(tempo_list)

    return average_tempo

def calculate_tempo_score(tempo_difference, threshold = 3):
    """
    두 곡간의 평균 템포 차이를 기반으로 점수를 빠르기 Score를 계산하는 함수
    
    Arg:
        tempo_difference: 템포 차이

    Returns:
        계산된 Score
    """
    excess_difference = max(0, tempo_difference - threshold) # 임계값을 초과한 템포 차이

    if tempo_difference <= threshold:
        # threshold 이하의 템포 차이는 최고 점수 100을 부여
        score = 100
    else:
        # threshold를 초과하는 경우 템포 차이에 따라 점수 부여
        # 차후에 가중치 생각해 볼 필요 있음 - 현재는 임시 로직
        score = max(0, 100 - min(3, excess_difference) * (tempo_difference - threshold))

    return f'{score:.3f}'

def calculate_tempo_level(score):
    """
    두 곡 간의 평균 템포 차이를 기반으로 Level을 계산하는 함수

    Arg:
        score: 현재 점수

    Returns:
        계산된 Level
    """
    score = float(score)  # 문자열에서 숫자로 변환
    level = int(min(10, round(score / 10))) # 최대 Level은 10
    return level

if __name__ == "__main__":
    # 두 MIDI 파일의 경로를 지정
    midi_file_path_1 = 'Prelude1.mid'
    midi_file_path_2 = 'Fugue1.mid'
    midi_file_path_3 = 'Fugue1.mid'
    midi_file_path_4 = 'Fugue3.mid'

    # 두 파일의 평균 템포를 추출
    average_tempo_1 = extract_tempo(midi_file_path_1)
    average_tempo_2 = extract_tempo(midi_file_path_2)
    average_tempo_3 = extract_tempo(midi_file_path_3)    
    average_tempo_4 = extract_tempo(midi_file_path_4)

    if average_tempo_1 is not None and average_tempo_2 is not None and average_tempo_3 is not None and average_tempo_3 is not None:
        # 평균 템포의 차이를 계산
        tempo_difference1 = abs(average_tempo_1 - average_tempo_2) # 다른 파일의 템포 평균
        tempo_difference2 = abs(average_tempo_3 - average_tempo_2) # 같은 파일의 템포 평균
        tempo_difference3 = abs(average_tempo_4 - average_tempo_3) # 다른 파일의 템포 평균
        
        # 템포 차이를 기반으로 점수 계산
        score1 = calculate_tempo_score(tempo_difference1) # 다른 파일로 테스트
        score2 = calculate_tempo_score(tempo_difference2) # 같은 파일로 테스트
        score3 = calculate_tempo_score(tempo_difference3) # 다른 파일로 테스트
        
        # 템포 차이에 따른 Level 계산 및 출력
        level_tempo1 = calculate_tempo_level(score1)
        level_tempo2 = calculate_tempo_level(score2)
        level_tempo3 = calculate_tempo_level(score3)
        
        print('Speed similarity scores of songs')
        print('--------------------------------------------------------------------------------------------')
        print(f'Average Tempo of {midi_file_path_1}: {average_tempo_1} BPM')
        print(f'Average Tempo of {midi_file_path_2}: {average_tempo_2} BPM')
        print(f'Average Tempo of {midi_file_path_4}: {average_tempo_3} BPM')
        print('--------------------------------------------------------------------------------------------')
        print(f'Tempo Difference Between {midi_file_path_1} and {midi_file_path_2}: {tempo_difference1} BPM')
        print(f'Tempo Similarirty Level testCase 1: Score: {score1} - Level: {level_tempo1}') # 다른 파일로 연주한 경우
        print('--------------------------------------------------------------------------------------------')
        print(f'Tempo Difference Between {midi_file_path_2} and {midi_file_path_3}: {tempo_difference2} BPM')
        print(f'Tempo Similarirty Level testCase 2: Score: {score2} - Level: {level_tempo2}') # 같은 파일로 연주한 경우
        print('--------------------------------------------------------------------------------------------')   
        print(f'Tempo Difference Between {midi_file_path_3} and {midi_file_path_4}: {tempo_difference3} BPM')
        print(f'Tempo Similarirty Level testCase 1: Score: {score3} - Level: {level_tempo3}') # 다른 파일로 연주한 경우
        print('--------------------------------------------------------------------------------------------')
    else:
        print('평균 템포를 계산할 수 없습니다.')

Speed similarity scores of songs
--------------------------------------------------------------------------------------------
Average Tempo of Prelude1.mid: 63.30000225000275 BPM
Average Tempo of Fugue1.mid: 53.12499922083624 BPM
Average Tempo of Fugue3.mid: 53.12499922083624 BPM
--------------------------------------------------------------------------------------------
Tempo Difference Between Prelude1.mid and Fugue1.mid: 10.175003029166511 BPM
Tempo Similarirty Level testCase 1: Score: 78.475 - Level: 8
--------------------------------------------------------------------------------------------
Tempo Difference Between Fugue1.mid and Fugue1.mid: 0.0 BPM
Tempo Similarirty Level testCase 2: Score: 100.000 - Level: 10
--------------------------------------------------------------------------------------------
Tempo Difference Between Fugue1.mid and Fugue3.mid: 18.43055086620948 BPM
Tempo Similarirty Level testCase 1: Score: 53.708 - Level: 5
--------------------------------------------

In [189]:
# 아래 코드는 바로 위 로직짜기 전에 작성한 연습 코드
import mido
from mido import MidiFile

def extract_tempo(midi_file):
    """
    MIDI 파일에서 템포 정보를 추출
    """
    mid = MidiFile(midi_file)

    for i, track in enumerate(mid.tracks):
        for msg in track:
            if msg.type == 'set_tempo':
                # 템포 메시지에서 마이크로초당 박자 속도를 추출
                tempo = mido.tempo2bpm(msg.tempo)
                return tempo

    # 템포 정보가 없는 경우 기본값으로 120 BPM을 반환.
    return 120

def tempo_difference(tempo1, tempo2):
    """
    두 템포 값 간의 유사도를 계산
    """
    # 예제로 간단하게 차이의 절대값을 반환
    return abs(tempo1 - tempo2)

if __name__ == "__main__":
    # 두 MIDI 파일의 경로를 지정.
    midi_file_path_1 = 'Prelude1.mid'
    midi_file_path_2 = 'Fugue1.mid'

    # 템포 정보를 추출
    tempo1 = extract_tempo(midi_file_path_1)
    tempo2 = extract_tempo(midi_file_path_2)

    # 템포 유사도를 계산
    difference = tempo_difference(tempo1, tempo2)

    print(f'Tempo of {midi_file_path_1}: {tempo1} BPM')
    print(f'Tempo of {midi_file_path_2}: {tempo2} BPM')
    print(f'Tempo Difference Between {midi_file_path_1} And {midi_file_path_2}: {difference} BPM')

Tempo of Prelude1.mid: 72.00002880001152 BPM
Tempo of Fugue1.mid: 64.0 BPM
Tempo Difference Between Prelude1.mid And Fugue1.mid: 8.00002880001152 BPM


# 5. 빠르기의 변화
- 추가적인 고민 필요

In [190]:
from mido import MidiFile, MidiTrack

def calculate_tempo_changes(midi_file_path):
    mid = MidiFile(midi_file_path)

    tempo_changes = []
    current_tempo = None

    for i, track in enumerate(mid.tracks):
        for msg in track:
            if msg.type == 'set_tempo':
                tempo = mido.tempo2bpm(msg.tempo)
                if current_tempo is not None and current_tempo != tempo:
                    time_in_seconds = mido.tick2second(msg.time, mid.ticks_per_beat, current_tempo)
                    tempo_changes.append((i, time_in_seconds, current_tempo, tempo))
                current_tempo = tempo

    return tempo_changes

def compare_tempo_changes(file1, file2):
    changes1 = calculate_tempo_changes(file1)
    changes2 = calculate_tempo_changes(file2)

    print(f"빠르기 변화 비교 ({file1} vs {file2}):")
    for i in range(min(len(changes1), len(changes2))):
        track1, time1, tempo1_start, tempo1_end = changes1[i]
        track2, time2, tempo2_start, tempo2_end = changes2[i]

        print(f"트랙 {track1 + 1} 및 {track2 + 1}, 시간: {time1:.2f}초 - {time2:.2f}초")
        print(f"{file1} 빠르기: {tempo1_start:.2f} BPM -> {tempo1_end:.2f} BPM")
        print(f"{file1} 빠르기 변화: {tempo1_end-tempo1_start:.2f} BPM") # 절댓값으로 하려면 abs() 사용
        print(f"{file2} 빠르기: {tempo2_start:.2f} BPM -> {tempo2_end:.2f} BPM")
        print(f"{file2} 빠르기 변화: {tempo2_end-tempo2_start:.2f} BPM") # 절댓값으로 하려면 abs() 사용
        print("---------------------------------------------------------------")

if __name__ == "__main__":
    midi_file1 = "Fugue3.mid"
    midi_file2 = "Fugue1.mid"
    midi_file3 = "Prelude1.mid"
    midi_file4 = "MIDI_sample.mid"

    #compare_tempo_changes(midi_file1, midi_file2)
    #compare_tempo_changes(midi_file1, midi_file3)
    compare_tempo_changes(midi_file1, midi_file3)

빠르기 변화 비교 (Fugue3.mid vs Prelude1.mid):
트랙 1 및 1, 시간: 0.02초 - 0.01초
Fugue3.mid 빠르기: 82.00 BPM -> 81.00 BPM
Fugue3.mid 빠르기 변화: -1.00 BPM
Prelude1.mid 빠르기: 72.00 BPM -> 70.00 BPM
Prelude1.mid 빠르기 변화: -2.00 BPM
---------------------------------------------------------------
트랙 1 및 1, 시간: 0.00초 - 0.00초
Fugue3.mid 빠르기: 81.00 BPM -> 79.00 BPM
Fugue3.mid 빠르기 변화: -2.00 BPM
Prelude1.mid 빠르기: 70.00 BPM -> 68.00 BPM
Prelude1.mid 빠르기 변화: -2.00 BPM
---------------------------------------------------------------
트랙 1 및 1, 시간: 0.00초 - 0.00초
Fugue3.mid 빠르기: 79.00 BPM -> 77.00 BPM
Fugue3.mid 빠르기 변화: -2.00 BPM
Prelude1.mid 빠르기: 68.00 BPM -> 66.00 BPM
Prelude1.mid 빠르기 변화: -2.00 BPM
---------------------------------------------------------------
트랙 1 및 1, 시간: 0.00초 - 0.00초
Fugue3.mid 빠르기: 77.00 BPM -> 72.00 BPM
Fugue3.mid 빠르기 변화: -5.00 BPM
Prelude1.mid 빠르기: 66.00 BPM -> 64.00 BPM
Prelude1.mid 빠르기 변화: -2.00 BPM
---------------------------------------------------------------
트랙 1 및 1, 시간: 0.00초 - 0.00초
Fugu

# 6. 붙임줄, 스타카토, 테누토, 늘임표

# 7. 악센트

# 8. 옥타브

In [191]:
from mido import MidiFile

def evaluate_octave_difference(input_path, target_path):
    input_mid = MidiFile(input_path)
    target_mid = MidiFile(target_path)

    # 각 곡의 옥타브 정보 추출
    input_octaves = [msg.note // 12 for track in input_mid.tracks for msg in track if msg.type == 'note_on']
    target_octaves = [msg.note // 12 for track in target_mid.tracks for msg in track if msg.type == 'note_on']

    # 두 곡의 옥타브 일치 여부를 평가하여 점수를 계산
    octave_matching_score = sum(1 for input_octave, target_octave in zip(input_octaves, target_octaves) if input_octave == target_octave)

    # 두 곡의 옥타브 일치 비율을 계산
    octave_matching_ratio = octave_matching_score / len(target_octaves)

    # 옥타브 일치 점수를 만점(100점) 기준으로 계산
    score = octave_matching_ratio * 100

    # 옥타브 정보 출력
    print(f"{input_path}의 옥타브 정보: {input_octaves}")
    print(f"{target_path}의 옥타브 정보: {target_octaves}")

    return score

def calculate_octave_level(score):
    """
    두 곡 간의 옥타브 평가 점수를 기반으로 Level을 계산하는 함수

    Arg:
        score: 현재 점수

    Returns:
        계산된 Level
    """
    score = float(score)  # 문자열에서 숫자로 변환
    level = int(min(10, round(score / 10))) # 최대 Level은 10
    return level

# MIDI 파일의 경로를 지정
input_path = 'Fugue1.mid'
target_path = 'Fugue3.mid'

# 평가 점수 계산
score = evaluate_octave_difference(input_path, target_path)
level = calculate_octave_level(score)

# 결과 출력
print(f"{input_path}과 {target_path}의 옥타브 평가 점수: {score:.2f} / 100")
print(f"{input_path}의 옥타브 평가 레벨: {level}")

Fugue1.mid의 옥타브 정보: [5, 5, 5, 6, 6, 6, 5, 6, 5, 6, 6, 6, 6, 5, 5, 5, 5, 6, 5, 6, 6, 6, 6, 6, 6, 6, 5, 6, 5, 6, 6, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 6, 5, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 5, 5, 6, 5, 5, 6, 5, 5, 6, 6, 6, 5, 5, 5, 5, 5, 6, 5, 5, 5, 5, 5, 5, 6, 6, 6, 6, 6, 5, 6, 5, 6, 5, 6, 5, 6, 5, 6, 5, 5, 5, 5, 5, 5, 6, 6, 6, 5, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 5, 5, 5, 6, 6, 6, 5, 6, 5, 6, 6, 6, 6, 5, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 5, 6, 5, 6, 6, 5, 6, 5, 6, 5, 5, 5, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 5, 5, 5, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 7, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 4, 4, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 6, 6, 6, 5, 6, 5, 6, 6, 6, 6, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5

In [192]:
input_path = 'Fugue1.mid'
target_path = 'Prelude1.mid'

# 평가 점수 계산
score = evaluate_octave_difference(input_path, target_path)
level = calculate_octave_level(score)

# 결과 출력
print(f"{input_path}과 {target_path}의 옥타브 평가 점수: {score:.2f} / 100")
print(f"{input_path}의 옥타브 평가 레벨: {level}")

Fugue1.mid의 옥타브 정보: [5, 5, 5, 6, 6, 6, 5, 6, 5, 6, 6, 6, 6, 5, 5, 5, 5, 6, 5, 6, 6, 6, 6, 6, 6, 6, 5, 6, 5, 6, 6, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 6, 5, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 5, 5, 6, 5, 5, 6, 5, 5, 6, 6, 6, 5, 5, 5, 5, 5, 6, 5, 5, 5, 5, 5, 5, 6, 6, 6, 6, 6, 5, 6, 5, 6, 5, 6, 5, 6, 5, 6, 5, 5, 5, 5, 5, 5, 6, 6, 6, 5, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 5, 5, 5, 6, 6, 6, 5, 6, 5, 6, 6, 6, 6, 5, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 5, 6, 5, 6, 6, 5, 6, 5, 6, 5, 5, 5, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 5, 5, 5, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 7, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 4, 4, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 6, 6, 6, 5, 6, 5, 6, 6, 6, 6, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5

In [193]:
input_path = 'sample_midi.mid'
target_path = 'sample_midi.mid'

# 평가 점수 계산
score = evaluate_octave_difference(input_path, target_path)
level = calculate_octave_level(score)

# 결과 출력
print(f"{input_path}과 {target_path}의 옥타브 평가 점수: {score:.2f} / 100")
print(f"{input_path}의 옥타브 평가 레벨: {level}")

sample_midi.mid의 옥타브 정보: [3, 4, 4, 5, 5, 3, 4, 4, 4, 5, 3, 4, 4, 4, 5, 3, 4, 4, 4, 5, 3, 4, 4, 5, 5, 3, 4, 4, 4, 5, 3, 4, 4, 5, 5, 3, 4, 4, 5, 5, 5, 3, 4, 4, 5, 5, 3, 4, 4, 4, 5, 3, 4, 4, 4, 5, 3, 4, 4, 4, 5, 3, 4, 4, 5, 5, 3, 4, 4, 4, 5, 3, 4, 4, 5, 5]
sample_midi.mid의 옥타브 정보: [3, 4, 4, 5, 5, 3, 4, 4, 4, 5, 3, 4, 4, 4, 5, 3, 4, 4, 4, 5, 3, 4, 4, 5, 5, 3, 4, 4, 4, 5, 3, 4, 4, 5, 5, 3, 4, 4, 5, 5, 5, 3, 4, 4, 5, 5, 3, 4, 4, 4, 5, 3, 4, 4, 4, 5, 3, 4, 4, 4, 5, 3, 4, 4, 5, 5, 3, 4, 4, 4, 5, 3, 4, 4, 5, 5]
sample_midi.mid과 sample_midi.mid의 옥타브 평가 점수: 100.00 / 100
sample_midi.mid의 옥타브 평가 레벨: 10


# 9. 꾸밈음, 반복 기호
- 추가 조사 필요

# 10. 리듬
- 추가 조사 필요

# 11. 페달링

In [72]:
file_path = "Fugue1.mid.mid"

mid = MidiFile(file_name)

In [73]:
mid.tracks[0]

MidiTrack([
  MetaMessage('time_signature', numerator=4, denominator=4, clocks_per_click=24, notated_32nd_notes_per_beat=8, time=0),
  MetaMessage('set_tempo', tempo=937500, time=0),
  MetaMessage('set_tempo', tempo=967742, time=24720),
  MetaMessage('set_tempo', tempo=1000000, time=240),
  MetaMessage('set_tempo', tempo=1052632, time=120),
  MetaMessage('set_tempo', tempo=1132075, time=120),
  MetaMessage('set_tempo', tempo=1200000, time=120),
  MetaMessage('set_tempo', tempo=1463415, time=64),
  MetaMessage('set_tempo', tempo=1578947, time=56),
  MetaMessage('end_of_track', time=0)])

In [215]:
from mido import MidiFile

def extract_pedal_and_note_durations(midi_file):
    mid = MidiFile(midi_file)

    note_durations = []  # 각 노트의 길이를 저장할 리스트
    pedal_states = []  # 페달 상태를 저장할 리스트 (페달이 눌려져 있는지 여부)
    control_values = []  # control_change 이벤트의 msg.control 값을 저장할 리스트

    current_time = 0  # 현재 시간 초기화
    current_pedal_state = 0  # 현재 페달 상태 초기화 (0은 페달 미작동)

    for i, track in enumerate(mid.tracks):
        for msg in track:
            current_time += msg.time  # 현재 시간을 누적

            if msg.type == 'note_on':
                note_start = current_time  # 노트 시작 시간
            elif msg.type == 'note_off':
                note_end = current_time  # 노트 종료 시간
                note_duration = note_end - note_start  # 노트 길이 계산
                note_durations.append(note_duration)
                pedal_states.append(current_pedal_state)  # 현재 페달 상태 저장
                control_values.append(None)  # 노트 이벤트에서는 control_change 이벤트가 없으므로 None 저장

            elif msg.type == 'control_change': # msg.contol == 64를 안해도 될까?
                # control_change 이벤트 발생 시 페달 상태 및 control 값을 저장
                current_pedal_state = 1 if msg.value > 0 else 0  # 0보다 크면 On, 아니면 Off로 처리
                note_durations.append(None)  # control_change 이벤트에서는 노트 길이가 없으므로 None 저장
                pedal_states.append(current_pedal_state)  # 현재 페달 상태 저장
                control_values.append(msg.control)  # control_change 이벤트의 msg.control 값 저장

    return note_durations, pedal_states

def evaluate_pedaling(input_path, target_path):
    # 두 곡의 페달 리스트 추출
    _, input_pedals = extract_pedal_and_note_durations(input_path)
    _, target_pedals = extract_pedal_and_note_durations(target_path)

    # 각각의 곡의 페달 상태 출력
    print(f"{input_path}의 페달 상태:", input_pedals)
    print(f"{target_path}의 페달 상태:", target_pedals)

    # 두 곡의 페달링 평가
    if len(input_pedals) == 0 or len(target_pedals) == 0:
        return 0.0  # 하나라도 곡이 페달 정보를 갖고 있지 않으면 평가 불가

    # 페달링이 일치하는 부분의 수를 계산
    matching_pedals = sum(1 for p1, p2 in zip(input_pedals, target_pedals) if p1 == p2)

    # 페달링 일치 비율을 계산
    pedal_matching_ratio = matching_pedals / len(input_pedals)

    # 페달링 평가 점수 반환 (만점 100점 기준)
    pedal_score = pedal_matching_ratio * 100

    return pedal_score

def calculate_pedal_level(score):
    """
    두 곡 간의 페달링 평가 점수를 기반으로 Level을 계산하는 함수

    Arg:
        score: 현재 점수

    Returns:
        계산된 Level
    """
    score = float(score)  # 문자열에서 숫자로 변환
    level = int(min(10, round(score / 10))) # 최대 Level은 10
    return level

# MIDI 파일의 경로를 지정
input_path = 'Fugue1.mid'
target_path = 'MIDI_sample.mid'

score = evaluate_pedaling(input_path, target_path)
level = calculate_pedal_level(score)

print(f"{target_path}에 대한 {input_path}의 페달링 평가 점수: {score:.2f} / 100")
print(f"{input_path}의 페달링 Level: {level}")

Fugue1.mid의 페달 상태: [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,

In [216]:
# MIDI 파일의 경로를 지정
input_path = 'Fugue3.mid'
target_path = 'Prelude1.mid'

score = evaluate_pedaling(input_path, target_path)
level = calculate_pedal_level(score)

print(f"{target_path}에 대한 {input_path}의 페달링 평가 점수: {score:.2f} / 100")
print(f"{input_path}의 페달링 Level: {level}")

Fugue3.mid의 페달 상태: [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,

In [219]:
# MIDI 파일의 경로를 지정
input_path = 'Fugue3.mid'
target_path = 'Fugue1.mid'

score = evaluate_pedaling(input_path, target_path)
level = calculate_pedal_level(score)

print(f"{target_path}에 대한 {input_path}의 페달링 평가 점수: {score:.2f} / 100")
print(f"{input_path}의 페달링 Level: {level}")

Fugue3.mid의 페달 상태: [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,

# 12. 노트의 길이 유사도 점수 구하기
- 조금 더 고민해봐야 할 것 같음

In [6]:
# 유사도 ?

def extract_note_durations(midi_file):
    """
    MIDI 파일에서 각 노트의 길이를 추출
    """
    mid = MidiFile(midi_file)

    note_durations = []  # 각 노트의 길이를 저장할 리스트

    current_time = 0  # 현재 시간 초기화

    for i, track in enumerate(mid.tracks):
        for msg in track:
            current_time += msg.time  # 현재 시간을 누적

            if msg.type == 'note_on':
                note_start = current_time  # 노트 시작 시간
            elif msg.type == 'note_off':
                note_end = current_time  # 노트 종료 시간
                note_duration = note_end - note_start  # 노트 길이 계산
                note_durations.append(note_duration)

    return note_durations

def adjust_note_lengths(note_durations_1, note_durations_2):
    """
    두 노트 길이 리스트의 길이를 맞추기
    """
    min_length = min(len(note_durations_1), len(note_durations_2))
    return note_durations_1[:min_length], note_durations_2[:min_length]

def calculate_note_duration_similarity(note_durations_1, note_durations_2, threshold = 10):
    """
    두 곡 간의 노트 길이 유사도 점수를 계산
    """
    # 노트 길이 리스트의 길이를 맞춤
    note_durations_1, note_durations_2 = adjust_note_lengths(note_durations_1, note_durations_2)

    # 노트 길이 차이의 절대값을 계산하여 유사도 점수 산출
    absolute_differences = [abs(a - b) for a, b in zip(note_durations_1, note_durations_2)]
    average_difference = sum(absolute_differences) / len(absolute_differences)
    if average_difference < threshold:
        return 100, note_durations_1, note_durations_2
    else:
        return average_difference % 100, note_durations_1, note_durations_2

if __name__ == "__main__":
    # MIDI 파일의 경로를 지정
    midi_file_path_1 = 'Fugue1.mid'
    midi_file_path_2 = 'Fugue3.mid'
    midi_file_path_3 = 'Prelude1.mid'
    

    # 각 MIDI 파일의 노트 길이를 추출
    note_durations_1 = extract_note_durations(midi_file_path_1)
    note_durations_2 = extract_note_durations(midi_file_path_2)
    note_durations_3 = extract_note_durations(midi_file_path_3)

    # 노트 길이 유사도 점수 계산 및 노트 길이 리스트 출력
    similarity_score, note_durations_1, note_durations_2 = calculate_note_duration_similarity(note_durations_1, note_durations_2)
    similarity_score2, note_durations_2, note_durations_3 = calculate_note_duration_similarity(note_durations_2, note_durations_3)
    similarity_score3, note_durations_3, note_durations_3 = calculate_note_duration_similarity(note_durations_3, note_durations_3)

    if similarity_score is not None and similarity_score2 is not None and similarity_score3 is not None:
        print(f'Note Duration Similarity Score: {similarity_score:.2f}%')
        print(f'Note Duration Similarity Score 2: {similarity_score2:.2f}%')
        print(f'Note Duration Similarity Score 3: {similarity_score3:.2f}%')
        print('-----------------------------------')
        print('Note Durations 1\n', note_durations_1)
        print('-----------------------------------')
        print('Note Durations 2\n', note_durations_2)
        print('-----------------------------------')
        print('Note Durations 3\n', note_durations_3)
    else:
        print('Error')


Note Duration Similarity Score: 71.50%
Note Duration Similarity Score 2: 8.58%
Note Duration Similarity Score 3: 100.00%
-----------------------------------
Note Durations 1
 [120, 120, 120, 180, 30, 2, 120, 120, 120, 170, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 120, 120, 120, 120, 60, 60, 60, 60, 170, 60, 60, 60, 60, 60, 60, 60, 60, 60, 480, 240, 120, 120, 120, 180, 30, 30, 120, 120, 120, 180, 60, 60, 60, 120, 180, 60, 60, 60, 480, 180, 60, 480, 240, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 120, 120, 120, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 120, 60, 60, 120, 120, 120, 240, 60, 60, 15, 15, 15, 15, 15, 15, 15, 15, 15, 45, 50, 240, 120, 120, 120, 180, 30, 2, 120, 120, 120, 120, 180, 30, 2, 120, 120, 120, 180, 60, 60, 60, 120, 120, 120, 120, 120, 60, 60, 60, 60, 60, 60, 60, 55, 5, 1, 30, 30, 2, 60, 60, 60, 60, 3, 1, 1, 1, 1, 100, 50, 148, 120, 120, 120, 147, 30, 2, 120, 120, 120, 170, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 3, 1, 

# 통합 테스트

Fugue1 vs Fugue3

|Pitch Accuracy|Speed Difference|Octave Accuracy|Pedaling Similarity|Note Duration Similarity|Mean|
|--|--|--|--|--|--|
|51.80%|55.00|17.70|100.00|71.50|59.2|

In [242]:
input_midi_path1 = 'Fugue1.mid'
target_midi_path1 = 'Fugue3.mid'
limit_offset = 5 # 정확도를 판단할 때 허용되는 피치 차이
accuracy1 = calculate_pitch_accuracy(input_midi_path1, target_midi_path1, limit_offset)
level1 = calculate_accuracy_level(accuracy1)  
print(f'Pitch Accuracy: {accuracy1:.2f}% - Level {level1}')

Pitch Accuracy: 51.80% - Level 5


In [243]:
average_tempo_1 = extract_tempo(input_midi_path1)
average_tempo_2 = extract_tempo(target_midi_path1)
tempo_difference1 = abs(average_tempo_1 - average_tempo_2)
score1 = calculate_tempo_score(tempo_difference1) # 다른 파일로 테스트        
level_tempo1 = calculate_tempo_level(score1)
print('Speed similarity scores of songs')
print('--------------------------------------------------------------------------------------------')
print(f'Average Tempo of {input_midi_path1}: {average_tempo_1} BPM')
print(f'Average Tempo of {input_midi_path2}: {average_tempo_2} BPM')
print('--------------------------------------------------------------------------------------------')
print(f'Tempo Difference Between {input_midi_path1} and {target_midi_path2}: {tempo_difference1} BPM')
print(f'Tempo Similarity Level testCase 1: Score: {score1} - Level: {level_tempo1}') # 다른 파일로 연주한 경우
print('--------------------------------------------------------------------------------------------')

Speed similarity scores of songs
--------------------------------------------------------------------------------------------
Average Tempo of Fugue1.mid: 64.0 BPM
Average Tempo of Prelude1.mid: 82.00003553334874 BPM
--------------------------------------------------------------------------------------------
Tempo Difference Between Fugue1.mid and Fugue1.mid: 18.000035533348736 BPM
Tempo Similarity Level testCase 1: Score: 55.000 - Level: 6
--------------------------------------------------------------------------------------------


In [239]:
score = evaluate_octave_difference(input_midi_path1, target_midi_path1)
level = calculate_octave_level(score)

# 결과 출력
print(f"{input_midi_path1}과 {target_midi_path1}의 옥타브 평가 점수: {score:.2f} / 100")
print(f"{input_midi_path1}의 옥타브 평가 레벨: {level}")

Fugue1.mid의 옥타브 정보: [5, 5, 5, 6, 6, 6, 5, 6, 5, 6, 6, 6, 6, 5, 5, 5, 5, 6, 5, 6, 6, 6, 6, 6, 6, 6, 5, 6, 5, 6, 6, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 6, 5, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 5, 5, 6, 5, 5, 6, 5, 5, 6, 6, 6, 5, 5, 5, 5, 5, 6, 5, 5, 5, 5, 5, 5, 6, 6, 6, 6, 6, 5, 6, 5, 6, 5, 6, 5, 6, 5, 6, 5, 5, 5, 5, 5, 5, 6, 6, 6, 5, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 5, 5, 5, 6, 6, 6, 5, 6, 5, 6, 6, 6, 6, 5, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 5, 6, 5, 6, 6, 5, 6, 5, 6, 5, 5, 5, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 5, 5, 5, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 7, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 4, 4, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 6, 6, 6, 5, 6, 5, 6, 6, 6, 6, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5

In [240]:
# MIDI 파일의 경로를 지정
input_midi_path1 = 'Fugue1.mid'
target_midi_path1 = 'Fugue3.mid'

score = evaluate_pedaling(input_midi_path1, target_midi_path1)
level = calculate_pedal_level(score)

print(f"{target_path}에 대한 {input_path}의 페달링 평가 점수: {score:.2f} / 100")
print(f"{input_path}의 페달링 Level: {level}")

Fugue1.mid의 페달 상태: [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,

In [241]:
# 각 MIDI 파일의 노트 길이를 추출
note_durations_1 = extract_note_durations(input_midi_path1)
note_durations_2 = extract_note_durations(target_midi_path1)
# 노트 길이 유사도 점수 계산 및 노트 길이 리스트 출력
similarity_score, note_durations_1, note_durations_2 = calculate_note_duration_similarity(note_durations_1, note_durations_2)

print(f'Note Duration Similarity Score: {similarity_score:.2f}%')
print('-----------------------------------')
print('Note Durations 1\n', note_durations_1)
print('Note Durations 1\n', note_durations_2)

Note Duration Similarity Score: 71.50%
-----------------------------------
Note Durations 1
 [120, 120, 120, 180, 30, 2, 120, 120, 120, 170, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 120, 120, 120, 120, 60, 60, 60, 60, 170, 60, 60, 60, 60, 60, 60, 60, 60, 60, 480, 240, 120, 120, 120, 180, 30, 30, 120, 120, 120, 180, 60, 60, 60, 120, 180, 60, 60, 60, 480, 180, 60, 480, 240, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 120, 120, 120, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 120, 60, 60, 120, 120, 120, 240, 60, 60, 15, 15, 15, 15, 15, 15, 15, 15, 15, 45, 50, 240, 120, 120, 120, 180, 30, 2, 120, 120, 120, 120, 180, 30, 2, 120, 120, 120, 180, 60, 60, 60, 120, 120, 120, 120, 120, 60, 60, 60, 60, 60, 60, 60, 55, 5, 1, 30, 30, 2, 60, 60, 60, 60, 3, 1, 1, 1, 1, 100, 50, 148, 120, 120, 120, 147, 30, 2, 120, 120, 120, 170, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 60, 3, 1, 1, 180, 120, 120, 120, 230, 120, 240, 120, 120, 120, 230, 120, 120, 120, 268, 38, 